openhands/core/message.py

```diff
@@ -84,33 +84,39 @@ def format_messages(
     for message in messages:
         content_parts = []
         role = 'user'
+        cache_prompt = False
 
         if isinstance(message, str) and message:
             content_parts.append(message)
         elif isinstance(message, dict):
             role = message.get('role', 'user')
             if 'content' in message and message['content']:
                 content_parts.append(message['content'])
+            cache_prompt = message.get('cache_prompt', False)
         elif isinstance(message, Message):
             role = message.role
             for content in message.content:
                 if isinstance(content, list):
                     for item in content:
                         if isinstance(item, TextContent) and item.text:
                             content_parts.append(item.text)
+                            cache_prompt |= item.cache_prompt
                 elif isinstance(content, TextContent) and content.text:
                     content_parts.append(content.text)
+                    cache_prompt |= content.cache_prompt
         else:
             logger.error(
                 f'>>> `message` is not a string, dict, or Message: {type(message)}'
             )
 
         if content_parts:
             content_str = '\n'.join(content_parts)
-            converted_messages.append(
-                {
-                    'role': role,
-                    'content': content_str,
-                }
-            )
+            formatted_message: Dict[str, Any] = {
+                'role': role,
+                'content': content_str,
+            }
+            if cache_prompt:
+                formatted_message['cache_control'] = {'type': 'ephemeral'}
```

Comment:
Can you please help me understand this, we are setting 'cache_control' on a message, not on a TextContext? I expected we need to do it on the TextContent or ImageContent, I seem to recall that the LLM will indeed respect it per item.

Reply:
What do you think about this: these are pydantic model classes, and they have each their own @model_serializer. When we used `model_dump` on the Message, it applied it to its component items, and we got the right format for free, so to speak. Now there seem to be some reasons why we don't rely on it anymore for the Messages, but could we still use it on TextContent and respectively ImageContent? It sets the appropriate cache_control if the `cache_prompt` attribute is set, I think?

Reply:
The code change is happening in a conversion method, which gets rid of the pydantic structure (like TextContent and  an extra type attribute), as that is an issue with several models, that error on that.

Reply:
@tobitege you mean that this [`type`](https://github.com/All-Hands-AI/OpenHands/blob/5100d12cea2cd35c30a22e25fbac376b72ed0981/openhands/core/message.py#L31) shouldn't be in the dict when vision is disabled or unsupported, but should be when vision is enabled and supported? Ugh. Then scratch that, it does affect TextContent.

I'd still appreciate an answer to the first question. We do something a little unusual [here](https://github.com/All-Hands-AI/OpenHands/blob/5100d12cea2cd35c30a22e25fbac376b72ed0981/agenthub/codeact_agent/codeact_agent.py#L289). This TextContent and thus the last Message it's part of, change next step, when the new message added to the list is now the last message and gets the reminder. I _think_ the way it worked with prompt caching is when the cache was set on the TextContent _before_ this last one.

Reply:
For e.g. Gemini (with vision) throws an error that the system message doesn't allow any extra attributes, which would be the `type` one, similarly it happens with models on Groq's API, iirc.

Yes, you're correct, we still have this `cache_prompt` on `TextContent`-level elsewhere:
https://github.com/All-Hands-AI/OpenHands/blob/5100d12cea2cd35c30a22e25fbac376b72ed0981/agenthub/codeact_agent/codeact_agent.py#L227-L238

Within this method here, the extra environmental hint would become part of a single-string (`messages` items on top) and not be its own message. 
Correct me if I'm wrong, but if vision were _enabled_ the problem would be with the pydantic serialization, not with the method here actually? 🤔

Reply:
Well, the CodeAct code on prompt caching worked in the past, using `model_dump()`.

> Within this method here, the extra environmental hint would become part of a single-string (messages items on top) and not be its own message.

No, it's its own TextContent object, added to a Message that has one other TextContent before it. So the result of that method you linked is: a Message with 2 TextContent objects, the first has `cache_prompt` and should be serialized with `cache_control`, the last is the reminder and should not have cache.

Why are we merging all TextContents in a single string? Sorry if I'm missing something, can't we send the dicts or something corresponding to the TextContent objects?

Reply:
Tobi, I'm a little confused here too:
> For e.g. Gemini (with vision) throws an error that the system message doesn't allow any extra attributes, which would be the `type` one, similarly it happens with models on Groq's API, iirc.

But as you said, if vision is enabled this method is not executed further, in this file at [line 80](https://github.com/All-Hands-AI/OpenHands/pull/3773/files#diff-1cdab1e3cbde6aeaac166359980171d0fe0c4c85688b5fe82cd38e249969cf23L80) it falls back to pydantic. So how did this method solve the problem of `type` for Gemini with vision? Or did you mean without vision? 🤔

Reply:
> So how did this method solve the problem of `type` for Gemini with vision? Or did you mean without vision? 🤔

Sorry, that is a typo indeed, I meant "without".